File: PE-common/src/main/java/nl/adaptivity/diagram/ItemCache.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Array;
import java.util.List;

/**
 * A cache implementation that allows drawing strategy related items to be stored.
 * This cache assumes that the strategy count is very low (1 or 2) and the item count
 * does not really change (and is low).
 *
 * @author Paul de Vrieze
 *
 */
public class ItemCache {

  @NotNull private DrawingStrategy<?,?,?>[] mStrategies = new DrawingStrategy<?,?,?>[0];
  @NotNull private Pen<?>[][] mPens = new Pen<?>[1][1];
  @NotNull private DiagramPath<?>[][] mPaths = new DiagramPath<?>[1][1];
  @NotNull
  @SuppressWarnings("unchecked")
  private List<DiagramPath<?>>[][] mPathLists = new List[1][1];

  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void setPen(final S strategy, final int index, final PEN_T pen) {
    final int strategyIdx = ensureStrategyIndex(strategy);
    mPens = ensureArrayLength(mPens, strategyIdx+1);
    Pen<?>[] sPens = mPens[strategyIdx];
    if (sPens==null) {
      mPens[strategyIdx] = (sPens = new Pen<?>[index+1]);
    } else {
      mPens[strategyIdx] = sPens = ensureArrayLength(sPens, index+1);
    }
    sPens[index] = pen;
  }

  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void setPath(final S strategy, final int index, final PATH_T path) {
    final int strategyIdx = ensureStrategyIndex(strategy);
    mPaths = ensureArrayLength(mPaths, strategyIdx+1);
    DiagramPath<?>[] sPaths = mPaths[strategyIdx];
    if (sPaths==null) {
      mPaths[strategyIdx] = sPaths = new DiagramPath<?>[index+1];
    } else {
      mPaths[strategyIdx] = sPaths = ensureArrayLength(sPaths, index+1);
    }
    sPaths[index] = path;
  }

  @SuppressWarnings({ "rawtypes", "unchecked" })
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void setPathList(final S strategy, final int index, final List<PATH_T> pathList) {
    final int strategyIdx = ensureStrategyIndex(strategy);
    mPathLists = ensureArrayLength(mPathLists, strategyIdx+1);
    List<DiagramPath<?>>[] sPathLists = mPathLists[strategyIdx];
    if (sPathLists==null) {
      mPathLists[strategyIdx] = sPathLists = new List[index+1];
    } else {
      mPathLists[strategyIdx] = sPathLists = ensureArrayLength(sPathLists, index+1);
    }
    sPathLists[index] = (List) pathList;
  }

  @NotNull
  private static <V> V[] ensureArrayLength(@NotNull final V[] array, final int length) {
    final int srcLen = Array.getLength(array);
    if (srcLen<length) {
      @SuppressWarnings("unchecked") final
      V[] newArray = (V[]) Array.newInstance(array.getClass().getComponentType(), length);
      System.arraycopy(array, 0, newArray, 0, srcLen);
      return newArray;
    }
    return array;
  }
//
//  private static <V> V[][] ensureArraysLength(V[][] array, int length) {
//    if (array.length<=length) {
//      @SuppressWarnings("unchecked")
//      V[][] newArray = (V[][]) Array.newInstance(array.getClass(), length, 0);
//      System.arraycopy(array, 0, newArray, 0, array.length);
//      return newArray;
//    }
//    return array;
//  }

  private <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> int getStrategyIndex(final S strategy) {
    int strategyIdx = -1;
    final int strategyLen = mStrategies.length;
    for (int i=0; i<strategyLen; ++i) {
      if (mStrategies[i]==strategy) {
        strategyIdx = i;
        break;
      }
    }
    return strategyIdx;
  }

  private <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> int ensureStrategyIndex(final S strategy) {
    int strategyIdx = getStrategyIndex(strategy);
    if (strategyIdx<0) {
      strategyIdx= mStrategies.length;
      if (mStrategies.length<=strategyIdx) {
        final DrawingStrategy<?,?,?>[] newStrategies = new DrawingStrategy<?,?,?>[strategyIdx+1];
        System.arraycopy(mStrategies, 0, newStrategies, 0, strategyIdx);
        newStrategies[strategyIdx]=strategy;
        mStrategies = newStrategies;
      } else {
        mStrategies[strategyIdx]=strategy;
      }
    }
    return strategyIdx;
  }

  @Nullable
  @SuppressWarnings("unchecked")
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> PEN_T getPen(final S strategy, final int index) {
    final int strategyIdx = getStrategyIndex(strategy);
    if(strategyIdx<0 || strategyIdx>= mPens.length || index>= mPens[strategyIdx].length) { return null; }
    return (PEN_T) mPens[strategyIdx][index];
  }

  @Nullable
  @SuppressWarnings("unchecked")
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> PATH_T getPath(final S strategy, final int index) {
    final int strategyIdx = getStrategyIndex(strategy);
    if(strategyIdx<0 || strategyIdx>= mPaths.length || index>= mPaths[strategyIdx].length) { return null; }
    return (PATH_T) mPaths[strategyIdx][index];
  }

  @Nullable
  @SuppressWarnings({ "unchecked", "rawtypes" })
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> List<PATH_T> getPathList(final S strategy, final int index) {
    final int strategyIdx = getStrategyIndex(strategy);
    if(strategyIdx<0 || strategyIdx>= mPathLists.length || index>= mPathLists[strategyIdx].length) { return null; }
    return (List) mPathLists[strategyIdx][index];
  }

  /**
   * Clear all the paths at the given index.
   * @param index The index of the paths to clear
   */
  public void clearPath(final int index) {
    for(final DiagramPath<?>[] lists: mPaths) {
      if (lists!=null && lists.length>index) {
        lists[index] = null;
      }
    }
  }

}

File: PE-common/common/src/main/kotlin/nl/adaptivity/diagram/ItemCache.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram

/**
 * A cache implementation that allows drawing strategy related items to be stored.
 * This cache assumes that the strategy count is very low (1 or 2) and the item count
 * does not really change (and is low).

 * @author Paul de Vrieze
 */
class ItemCache {

  private var strategies = arrayOfNulls<DrawingStrategy<*, *, *>>(0)
  private var pens = arrayOf<Array<Pen<*>?>?>(arrayOfNulls<Pen<*>>(1))
  private var paths = arrayOf<Array<DiagramPath<*>?>?>(arrayOfNulls<DiagramPath<*>>(1))
  private var pathLists = arrayOf<Array<List<*>?>?>(arrayOfNulls<List<*>>(1))

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> setPen(strategy: S, index: Int, pen: PEN_T) {
    setPen(ensureStrategyIndex(strategy), index, pen)
  }

  private fun <PEN_T : Pen<PEN_T>> setPen(strategyIdx: Int, penIdx: Int, pen: PEN_T) {
    pens = pens.ensureArrayLength(strategyIdx + 1)

    val strategyPens = pens[strategyIdx].ensureArrayLength(penIdx + 1)

    strategyPens[penIdx] = pen
    pens[strategyIdx] = strategyPens
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> setPath(strategy: S, index: Int, path: PATH_T) {
    setPath(ensureStrategyIndex(strategy), index, path)
  }

  private fun <PATH_T : DiagramPath<PATH_T>> setPath(strategyIdx: Int, pathIdx: Int, path: PATH_T) {
    if (strategyIdx<0) {
      setPath(strategies.size, pathIdx, path)
    } else {
      paths = paths.ensureArrayLength(strategyIdx + 1)
      val sPaths = paths[strategyIdx].ensureArrayLength(pathIdx + 1)

      sPaths[pathIdx] = path
      paths[strategyIdx] = sPaths
    }
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> setPathList(strategy: S, index: Int, pathList: List<PATH_T>) {
    setPathList(ensureStrategyIndex(strategy), index, pathList)
  }

  private fun <PATH_T : DiagramPath<PATH_T>> setPathList(strategyIdx: Int, pathListIdx: Int, pathList: List<PATH_T>) {
    if (strategyIdx<0) {
      setPathList(strategies.size, pathListIdx, pathList)
    } else {
      pathLists = pathLists.ensureArrayLength(strategyIdx + 1)
      val sPathLists = pathLists[strategyIdx].ensureArrayLength(pathListIdx + 1)
      pathLists[strategyIdx] = sPathLists
      sPathLists[pathListIdx] = pathList as List<*>
    }
  }

  inline private fun <reified V> Array<V?>?.ensureArrayLength(length: Int) = when {
    this==null -> arrayOfNulls<V>(length)
    else -> this.ensureArrayLengthNotNull(length)
  }

  private fun <V> Array<V?>.ensureArrayLengthNotNull(length: Int): Array<V?> {
    val array = this
    val srcLen = array.size
    if (srcLen < length) {
        return array + arrayOfNulls(length - srcLen)
    }
    return array
  }
  //
  //  private static <V> V[][] ensureArraysLength(V[][] array, int length) {
  //    if (array.length<=length) {
  //      @SuppressWarnings("unchecked")
  //      V[][] newArray = (V[][]) Array.newInstance(array.getClass(), length, 0);
  //      System.arraycopy(array, 0, newArray, 0, array.length);
  //      return newArray;
  //    }
  //    return array;
  //  }

  private fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getStrategyIndex(strategy: S): Int {
    return strategies.indexOfFirst { it===strategy }
  }

  private fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> ensureStrategyIndex(strategy: S): Int {
    var strategyIdx = getStrategyIndex(strategy)
    if (strategyIdx < 0) {
      strategyIdx = strategies.size
      val newStrategies = strategies.ensureArrayLengthNotNull(strategyIdx+1)
      newStrategies[strategyIdx] = strategy
      strategies = newStrategies
    }
    return strategyIdx
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPen(strategy: S, index: Int, alternate: PEN_T.()->Unit): PEN_T {
    val strategyIdx = getStrategyIndex(strategy)
    return getPen<PEN_T>(strategyIdx, index) ?: strategy.newPen().apply { alternate(); setPen(strategyIdx, index, this) }
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPen(strategy: S, index: Int): PEN_T? {
    val strategyIdx = getStrategyIndex(strategy)

    return getPen<PEN_T>(strategyIdx, index)
  }

  private fun <PEN_T : Pen<PEN_T>> getPen(strategyIdx: Int, penIdx: Int): PEN_T? {
    if (strategyIdx < 0 || strategyIdx >= pens.size || penIdx >= pens[strategyIdx]!!.size) {
      return null
    }
    return pens[strategyIdx]!![penIdx] as PEN_T?
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPath(strategy: S, index: Int, alternate:PATH_T.()->Unit): PATH_T {
    val strategyIdx = getStrategyIndex(strategy)
    return getPath<S, PEN_T, PATH_T>(strategy, index) ?: strategy.newPath().apply { alternate(); setPath(strategyIdx, index, this) }
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPath(strategy: S, index: Int): PATH_T? {
    val strategyIdx = getStrategyIndex(strategy)
    return getPath<PATH_T>(strategyIdx, index)
  }

  private fun <PATH_T : DiagramPath<PATH_T>> getPath(strategyIdx: Int, pathIdx: Int): PATH_T? {
    if (strategyIdx < 0 || strategyIdx >= paths.size || pathIdx >= paths[strategyIdx]!!.size) {
      return null
    }
    return paths[strategyIdx]!![pathIdx] as PATH_T?
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPathList(strategy: S, index: Int, alternate: (S)->List<PATH_T>): List<PATH_T> {
    val strategyIdx = getStrategyIndex(strategy)
    return getPathList<S, PEN_T, PATH_T>(strategyIdx, index) ?: alternate(strategy).apply { setPathList(strategyIdx, index, this) }
  }

  fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPathList(strategy: S, index: Int): List<PATH_T>? {
    val strategyIdx = getStrategyIndex(strategy)
    return getPathList<S, PEN_T, PATH_T>(strategyIdx, index)
  }

  private fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> getPathList(strategyIdx: Int, pathListIdx: Int): List<PATH_T>? {
    if (strategyIdx < 0 || strategyIdx >= pathLists.size || pathListIdx >= pathLists[strategyIdx]!!.size) {
      return null
    }
    return pathLists[strategyIdx]!![pathListIdx] as List<PATH_T>?
  }

  /**
   * Clear all the paths at the given index.
   * @param index The index of the paths to clear
   */
  fun clearPath(index: Int) {
    for (lists in paths) {
      if (lists != null && lists!!.size > index) {
        lists[index] = null
      }
    }
  }

}
